packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,186 @@
+          const runState = await this.getSingleState(runId, await this._accessToken);
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = this.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  public clearJob(reportId: string) {
+    this._reportRunIds.delete(reportId);
+  }
+
+  private getFinalState(states: ExtractorState[]) {
+    if (states.includes(ExtractorState.Failed))
+      return ExtractionStates.Failed;
+
+    if (states.includes(ExtractorState.Queued))
+      return ExtractionStates.Queued;
+
+    if (states.includes(ExtractorState.Running))
+      return ExtractionStates.Running;
+
+    if (states.includes(ExtractorState.Succeeded))
+      return ExtractionStates.Succeeded;
+
+    return ExtractionStates.Failed;
+  }
+
+  private async getSingleState(runId: string, accessToken: string): Promise<ExtractorState> {
+    try {
+      const response = await this._extractionClientApi.getExtractionStatus(accessToken, runId);
+      return response.state;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return ExtractorState.Failed;
+  }
+
+  public async startJobs(reportIds: string[]) {
+    const extractionByIModel = new Map<string, string>();
+    for (const reportId of reportIds) {
+      const iModels = (await this.fetchReportMappings(reportId)).map((m) => m.imodelId);
+      const uniqueIModels = new Set(iModels);
+
+      const runs: string[] = [];
+      for (const iModelId of uniqueIModels) {
+        const extraction = extractionByIModel.get(iModelId);
+        if (extraction) {
+          runs.push(extraction);
+        } else {
+          const run = await this.runExtraction(iModelId, await this._accessToken);
+          if (run) {
+            runs.push(run.id);
+            extractionByIModel.set(iModelId, run.id);
+          }
+        }
+      }
+      this._reportRunIds.set(reportId, runs);
+    }
+  }
+
+  private async runExtraction(iModelId: string, accessToken: string) {
+    try {
+      const response = await this._extractionClientApi.runExtraction(
+        accessToken,
+        iModelId
+      );
+      return response;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return undefined;
+  }
+
+  private async fetchReportMappings(
+    reportId: string
+  ): Promise<ReportMappingAndMapping[]> {
+    try {
+      const reportMappings = await this._reportsClientApi.getReportMappings(
+        await this._accessToken,
+        reportId
+      );
+      const authorization =
+        AccessTokenAdapter.toAuthorizationCallback(await this._accessToken);
+
+      const reportMappingsAndMapping = await Promise.all(
+        reportMappings.map(async (reportMapping) => {
+          const iModelId = reportMapping.imodelId;
+          let iModelName = "";
+          const mapping = await this._mappingsClientApi.getMapping(
```

Comment:
A single `startJobs` call for multiple reports might result in multiple `getMapping` requests for the same mapping.

Reply:
`startJobs` has been refactored. `getMapping` will not be called. Only unique `iModels` will be used